### ISSUE TITLE
[Bug] buildx failed with: ERROR: failed to solve: failed to read dockerfile: failed to load cache key: repository does not contain ref

### ISSUE THREAD

**Issue by chandanpg** [docker/build-push-action#1222]
### Contributing guidelines

- [X] I've read the [contributing guidelines](https://github.com/docker/build-push-action/blob/master/.github/CONTRIBUTING.md) and wholeheartedly agree

### I've found a bug, and:

- [X] The documentation does not mention anything about my problem
- [X] There are no open or closed issues that are related to my problem

### Description

I noticed recently that we are getting error in CICD when we are calling this action. This error is coming when you raise a Pull request and you merge it immediately. You Docker Build Process(which is calling this action) during Pull Request would fail with below error.

Error: buildx failed with: ERROR: failed to solve: failed to read dockerfile: failed to load cache key: repository does not contain ref refs/pull/9/merge, output: ""

### Expected behaviour

Job should go to success even when your Pull Request is merged/closed. 

### Actual behaviour

Job is failing when Pull Request is merged immediately 

### YAML workflow

```yaml
- name: Build
      id: build
      uses: docker/build-push-action@v6
      with:
        push: ${{ github.event_name != 'pull_request' }}
        network: host
        file: Dockerfile
        provenance: false
        github-token: ${{ inputs.TOKEN }}
        secret-files: |
            npmrc=${{ env.NPMRC_PATH }}
        secrets: |
            "TOKEN=${{ inputs.TOKEN }}"
        build-args: |
            "GHA_RUN_NUMBER=${{ inputs.GHA_RUN_NUMBER }}"
            ${{ inputs.DOCKER_ARGS }}
        tags: |
            "us-east4-docker.pkg.dev/${{ inputs.GCP_PROJECT_ID }}/${{ env.REPO_NAME }}/${{ inputs.IMAGE_NAME }}:latest"
```

**Comment by crazy-max** [comment]
Can you provide workflow logs and [BuildKit logs](https://docs.docker.com/build/ci/github-actions/configure-builder/#buildkit-container-logs) please?

**Comment by chandanpg** [comment]
Hi, I have attached the logs as well[](url)

**Comment by chandanpg** [comment]
[3_CI _ Build.zip](https://github.com/user-attachments/files/16838237/3_CI._.Build.zip)

**Comment by crazy-max** [comment]
> This error is coming when you raise a Pull request and you merge it immediately

I'm not quite sure what that means. Does the PR got merged but workflow is still running? In that case it might be that the ref get removed when merged and therefore BuildKit is not able to fetch?

Looking at the workflow logs, I see the checkout action uses another ref:

```
2024-09-02T07:47:05.1328165Z ##[group]Fetching the repository
2024-09-02T07:47:05.1336097Z [command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +c5890f87dd023da50edd68c661f16134495e8cb5:refs/remotes/pull/9/merge
2024-09-02T07:47:05.3928376Z From https://github.com/my-org/my-repo
2024-09-02T07:47:05.3930879Z  * [new ref]         c5890f87dd023da50edd68c661f16134495e8cb5 -> pull/9/merge
...
2024-09-02T07:47:05.4042568Z ::group::Checking out the ref
2024-09-02T07:47:05.4043346Z ##[group]Checking out the ref
2024-09-02T07:47:05.4044939Z [command]/usr/bin/git checkout --progress --force refs/remotes/pull/9/merge
2024-09-02T07:47:05.4181348Z Note: switching to 'refs/remotes/pull/9/merge'.
```

I wonder if we should do something similar. I'm taking a look.

**Comment by crazy-max** [comment]
In the meantime you can use the path context: https://github.com/docker/build-push-action/?tab=readme-ov-file#path-context

```yaml
    - name: Build
      id: build
      uses: docker/build-push-action@v6
      with:
        context: .
        push: ${{ github.event_name != 'pull_request' }}
        network: host
        file: Dockerfile
        provenance: false
        github-token: ${{ inputs.TOKEN }}
        secret-files: |
            npmrc=${{ env.NPMRC_PATH }}
        secrets: |
            "TOKEN=${{ inputs.TOKEN }}"
        build-args: |
            "GHA_RUN_NUMBER=${{ inputs.GHA_RUN_NUMBER }}"
            ${{ inputs.DOCKER_ARGS }}
        tags: |
            "us-east4-docker.pkg.dev/${{ inputs.GCP_PROJECT_ID }}/${{ env.REPO_NAME }}/${{ inputs.IMAGE_NAME }}:latest"
```

**Comment by chandanpg** [comment]
> > This error is coming when you raise a Pull request and you merge it immediately
> 
> I'm not quite sure what that means. Does the PR got merged but workflow is still running? In that case it might be that the ref get removed when merged and therefore BuildKit is not able to fetch?
> 
> Looking at the workflow logs, I see the checkout action uses another ref:
> 
> ```
> 2024-09-02T07:47:05.1328165Z ##[group]Fetching the repository
> 2024-09-02T07:47:05.1336097Z [command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +c5890f87dd023da50edd68c661f16134495e8cb5:refs/remotes/pull/9/merge
> 2024-09-02T07:47:05.3928376Z From https://github.com/my-org/my-repo
> 2024-09-02T07:47:05.3930879Z  * [new ref]         c5890f87dd023da50edd68c661f16134495e8cb5 -> pull/9/merge
> ...
> 2024-09-02T07:47:05.4042568Z ::group::Checking out the ref
> 2024-09-02T07:47:05.4043346Z ##[group]Checking out the ref
> 2024-09-02T07:47:05.4044939Z [command]/usr/bin/git checkout --progress --force refs/remotes/pull/9/merge
> 2024-09-02T07:47:05.4181348Z Note: switching to 'refs/remotes/pull/9/merge'.
> ```
> 
> I wonder if we should do something similar. I'm taking a look.

Yes correct. PR is getting merged immediately after opening the Pull request. When Pull Request is merged, Workflow for Pull request is still running .

**Comment by chandanpg** [comment]
@crazy-max  Can you please prioritize this as it impacting many users

**Comment by crazy-max** [comment]
So after investigating it seems the `actions/checkout` uses a different refspec for pull requests:

```
2024-09-02T07:47:05.1328165Z ##[group]Fetching the repository
2024-09-02T07:47:05.1336097Z [command]/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 origin +c5890f87dd023da50edd68c661f16134495e8cb5:refs/remotes/pull/9/merge
2024-09-02T07:47:05.3928376Z From https://github.com/my-org/my-repo
2024-09-02T07:47:05.3930879Z  * [new ref]         c5890f87dd023da50edd68c661f16134495e8cb5 -> pull/9/merge
...
2024-09-02T07:47:05.4042568Z ::group::Checking out the ref
2024-09-02T07:47:05.4043346Z ##[group]Checking out the ref
2024-09-02T07:47:05.4044939Z [command]/usr/bin/git checkout --progress --force refs/remotes/pull/9/merge
2024-09-02T07:47:05.4181348Z Note: switching to 'refs/remotes/pull/9/merge'.
```

It uses a refspec to fetch the merged commit (`c5890f87dd023da50edd68c661f16134495e8cb5`) and then updates the local reference `refs/remotes/pull/9/merge` with this commit. The + sign forces the update even if it results in a non-fast-forward update.

> Can you please prioritize this as it impacting many users

We only have a single report from you afaik but please share info if there are more people impacted.

Do you have limitations using the path context vs git context in the meantime as show in https://github.com/docker/build-push-action/issues/1222#issuecomment-2324893530?

**Comment by crazy-max** [comment]
> So after investigating

Was testing on private repo my bad so here is other investigation with a public repo: https://github.com/crazy-max/buildx-buildkit-tests/actions/runs/11073531258

The `actions/checkout` uses a different refspec for pull requests: https://github.com/crazy-max/buildx-buildkit-tests/actions/runs/11073531258/job/30770257801#step:4:53

![image](https://github.com/user-attachments/assets/6d869a28-fe1a-43f0-990a-d7ef33a5cda0)

It uses a refspec `+3231c3fa12afbe81fbb290db15a994004a53ec4a:refs/remotes/pull/1/merge` to fetch the merged commit (`3231c3fa12afbe81fbb290db15a994004a53ec4a`) and then updates the local reference `refs/remotes/pull/1/merge` with this commit. The `+` sign forces the update even if it results in a non-fast-forward update.

In our case we are using the provided `github.ref` from github context, here `refs/pull/1/merge`: https://github.com/crazy-max/buildx-buildkit-tests/actions/runs/11073531258/job/30770258123#step:4:137

![image](https://github.com/user-attachments/assets/90566cde-ae6a-4f4d-b855-44d5de9488fe)

I tried fetching directly from commit sha as well but unable to read tree: https://github.com/crazy-max/buildx-buildkit-tests/actions/runs/11073531258/job/30770257444#step:4:138

![image](https://github.com/user-attachments/assets/0a497454-7c8c-4533-9a0c-9d6339cdf572)

I'm not quite sure how we can handle that in Git source https://github.com/moby/buildkit/blob/master/source/git/source.go atm :thinking: Might require changes around https://github.com/moby/buildkit/blob/185751bdc010f1f8bf1192d6e9b4ec19552d5741/source/git/source.go#L368-L405

What seems to work is using the ref that points to the latest commit on the source branch of the pull request with `refs/pull/1/head`: https://github.com/crazy-max/buildx-buildkit-tests/actions/runs/11073531258/job/30770257072#step:5:140

![image](https://github.com/user-attachments/assets/5c01cf0a-ea8d-4de6-b537-6638c2cac0b3)

**Comment by crazy-max** [comment]
@chandanpg Should be fixed with latest release
